src/arena.c

```diff
@@ -491,6 +491,20 @@ arena_extent_ralloc_large_expand(tsdn_t *tsdn, arena_t *arena, extent_t *extent,
 	arena_nactive_add(arena, udiff >> LG_PAGE);
 }
 
+static ssize_t
+arena_decay_time_read(arena_t *arena) {
+	arena_decay_time_t dt;
+	dt.u = atomic_read_zu(&arena->decay.time.u);
+	return dt.s;
```

Comment:
The C semantics are only deceptively nice, since they're too vague to be useful in interesting cases. There's a note that says `u.a = x; y = u.b;` does the obvious thing, but it's also fairly clear that union magic doesn't apply to uses that merely have possibly aliasing pointers to union members, and there's a bunch of area in between those extremes.

```
void func_a(type_a *);
void func_b(type_b *);
void func_a_b(type_a *, type_b *);
typedef union {
  type_a a;
  type_b b;
} U;
void probably_safe() {
  U u;
  func_a(&u.a);
  func_b(&u.b);
}
void risky() {
  U u;
  type_a *a_ptr = &u.a;
  type_b *b_ptr = &u.b;
  func_a(a_ptr);
  func_b(b_ptr);
}
```

So really the question has to be about acceptable levels of compiler risk instead of a binary "UB: yes/no" unless we're willing to color entirely within the lines (which I suspect will be impossible in a malloc).